### PR TITLE
Adds support for ANCM in process

### PR DESCRIPTION
**Pull request by jkotalik**
Initial implementation of ANCM in process' managed component. This will be the only large feature implementation, with the rest coming in smaller feature sets. 
Notes:
- Uses a mix of HttpSys's native structs and Kestrel's control flow
- Exposes a pInvoke layer to call into ANCM via the NativeMethods class
- Adds some functional tests for general cases

There are many features that still need to be implemented/verified. I will make corresponding issues after this is merged in:
- Authentication and Windows Auth
- Request and Response headers verification
- Performance in general
- Pathbase and Path identification
- Versioning between IISIntegration and ANCM. 
- Add more functional tests
- Mock ANCM for unit tests (something like [libuv mocking](https://github.com/aspnet/KestrelHttpServer/blob/dev/test/Kestrel.Transport.Libuv.Tests/TestHelpers/MockLibuv.cs)?)
- Main server control flow verification
- Verify System.IO.Pipeline changes didn't cause regression. 

cc/ @Eilon @muratg @pan-wang @shirhatti

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.FeatureCollection.cs [hunk 1]
```diff
@@ -0,0 +1,270 @@
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.FeatureCollection.cs]: Taking from Kestrel. We will see if these features can be combined in some way in the future.

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.Features.cs [hunk 2]
```diff
@@ -0,0 +1,322 @@
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.Features.cs]: Again taken from Kestrel.

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 3]
```diff
@@ -0,0 +1,872 @@
...
+
+            // We should be sending the headers here as there are many responses that don't have status codes.
+            _pinnedHeaders = SerializeHeaders(pHttpResponse);
+        }
+
+        // From HttpSys, except does not write to response
+        private unsafe List<GCHandle> SerializeHeaders(HttpApiTypes.HTTP_RESPONSE_V2* pHttpResponse)
+        {
+            // TODO make the header collection readonly to stop modifications.
+            HttpApiTypes.HTTP_UNKNOWN_HEADER[] unknownHeaders = null;
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: From HttpSys, except we don't send the header response here. Could maybe consider some code sharing in the future, however we need to be careful about buffer management.

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 4]
```diff
@@ -0,0 +1,872 @@
...
+                {
+                    Output.Reader.Advance(consumed);
+                }
+            }
+            Output.Reader.Complete();
+        }
+
+        private unsafe IISAwaitable WriteAsync(ReadableBuffer buffer)
+        {
+            var fCompletionExpected = false;
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: We need to be careful about this writing logic.

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting.Server;
+using System.Threading;
+using System.IO.Pipelines;
+
+namespace Microsoft.AspNetCore.Server.IIS
+{
+    internal class IISHttpContextOfT<TContext> : HttpProtocol
+    {
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs]: Taken from kestrel, probably should remove commented out code here.

File: src/Microsoft.AspNetCore.Server.IIS/NativeMethods.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        private const string AspNetCoreModuleDll = "aspnetcore.dll";
+
+        public enum REQUEST_NOTIFICATION_STATUS
+        {
+            RQ_NOTIFICATION_CONTINUE,
+            RQ_NOTIFICATION_PENDING,
+            RQ_NOTIFICATION_FINISH_REQUEST
+        }
+
+        public delegate REQUEST_NOTIFICATION_STATUS PFN_REQUEST_HANDLER(IntPtr pHttpContext, IntPtr pvRequestContext);
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IIS/NativeMethods.cs]: These need to stay in sync with ANCM. We will also need to expose a version api check here.

File: test/IISIntegration.IISServerFunctionalTests/FeatureCollectionTests.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+using Microsoft.Extensions.Logging.Testing;
+using Xunit;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+
+namespace Microsoft.AspNetCore.Server.IIS.FunctionalTests
+{
+    public class FeatureCollectionTest : LoggedTest
+    {
+        public FeatureCollectionTest(ITestOutputHelper output) : base(output)
```
Comment (jkotalik) [inline test/IISIntegration.IISServerFunctionalTests/FeatureCollectionTests.cs]: Going to disable tests until is fixed (or we build ANCM from source) https://github.com/aspnet/IISIntegration/issues/424

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs (hunk 5 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs]: Also should consider adding corresponding options class.

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+            var context = default(TContext);
+
+            try
+            {
+                context = _application.CreateContext(this);
+
+                await _application.ProcessRequestAsync(context);
+                //if (Volatile.Read(ref _requestAborted) == 0)
+                //{
+                //    VerifyResponseContentLength();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs]: TODO

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+            if (Volatile.Read(ref _requestAborted) == 0)
+            {
+                if (HasResponseStarted)
+                {
+                    // If the response has already started, call ProduceEnd() before
+                    // consuming the rest of the request body to prevent
+                    // delaying clients waiting for the chunk terminator:
+                    //
+                    // https://github.com/dotnet/corefx/issues/17330#issuecomment-288248663
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs]: I don't think any of these comments are relevent. clean up

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+                    //
+                    // ProduceEnd() must be called before _application.DisposeContext(), to ensure
+                    // HttpContext.Response.StatusCode is correctly set when
+                    // IHttpContextFactory.Dispose(HttpContext) is called.
+                    await ProduceEnd();
+                }
+
+                //// ForZeroContentLength does not complete the reader nor the writer
+                //if (!messageBody.IsEmpty && _keepAlive)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs]: remove

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+                //{
+                //    // Finish reading the request body in case the app did not.
+                //    TimeoutControl.SetTimeout(Constants.RequestBodyDrainTimeout.Ticks, TimeoutAction.SendTimeoutResponse);
+                //    await messageBody.ConsumeAsync();
+                //    TimeoutControl.CancelTimeout();
+                //}
+
+                if (!HasResponseStarted)
+                {
+                    await ProduceEnd();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs]: combine with earlier call?

File: build/repo.props [hunk 12]
```diff
@@ -1,6 +1,7 @@
-﻿<Project>
+<Project>
   <ItemGroup>
     <ExcludeFromTest Include="$(RepositoryRoot)test\TestSites\*.csproj" />
+    <ExcludeFromTest Include="$(RepositoryRoot)test\IISTestSite\*.csproj" />
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply build/repo.props]: Program.cs differences. Also distinctly separate scenarios. I can put them into the same folder or if we decided to merge with UseIISIntegration.

File: samples/NativeIISSample/Startup.cs [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+                    await context.Response.WriteAsync(key + ": " + value + Environment.NewLine);
+                }
+                await context.Response.WriteAsync(Environment.NewLine);
+            });
+        }
+
+        public static void Main(string[] args)
+        {
+            var host = new WebHostBuilder()
+                .UseNativeIIS()
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply samples/NativeIISSample/Startup.cs]: For now I will leave it separate. I'll file a bug to consider changing this.

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.FeatureCollection.cs [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+    {
+        // NOTE: When feature interfaces are added to or removed from this Frame class implementation,
+        // then the list of `implementedFeatures` in the generated code project MUST also be updated.
+        // See also: tools/Microsoft.AspNetCore.Server.Kestrel.GeneratedCode/FrameFeatureCollection.cs
+
+        private int _featureRevision;
+
+        private List<KeyValuePair<Type, object>> MaybeExtra;
+
+        public void ResetFeatureCollection()
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.FeatureCollection.cs]: Currently used [here](https://github.com/aspnet/IISIntegration/pull/425/files/968c85fe745f08c4a07f53bcefb71dc747c441cb#diff-2fa3812e6254f468964ecdb35d0ae1ebR137)

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.FeatureCollection.cs [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+
+        string IHttpRequestFeature.Method
+        {
+            get => Method;
+            set => Method = value;
+        }
+
+        string IHttpRequestFeature.PathBase
+        {
+            get => PathBase ?? "";
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.FeatureCollection.cs]: I'm currently in favor of removing the checks entirely.

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 16]
```diff
@@ -0,0 +1,870 @@
...
+        private IISAwaitable _flushOperation = new IISAwaitable();
+
+        private TaskCompletionSource<object> _upgradeTcs;
+
+        protected Task _readingTask;
+        protected Task _writingTask;
+
+        protected int _requestAborted;
+
+        private static bool _upgradeAvailable = (Environment.OSVersion.Version >= new Version(6, 2));
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: I remember Pan mentioning that there was no way to check in ANCM if WebSockets is installed. Is there another check we can do?

[the remark this answers is not in the document]
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: That wasn't my question. My question was does it matter? Can you do an Upgrade even if they aren't enabled? Note per http.sys you have to at least set the upgrade header to websockets for the Upgrade to succeed.

**Comment by jkotalik** [comment]
:up: :date: I made many corresponding issues to track issues. 
I was considering adding issues for:
- Correctness of Reading and Writing to and from ANCM (no out of order R/W)
- More Functional Tests

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 17]
```diff
@@ -0,0 +1,875 @@
...
+        private IISAwaitable _flushOperation = new IISAwaitable();
+
+        private TaskCompletionSource<object> _upgradeTcs;
+
+        protected Task _readingTask;
+        protected Task _writingTask;
+
+        protected int _requestAborted;
+
+
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: cleanup

File: src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs [hunk 18]
```diff
@@ -0,0 +1,875 @@
...
+            HttpApiTypes.HTTP_RESPONSE_V2* pHttpResponse = NativeMethods.http_get_raw_response(_pHttpContext);
+
+            // We should be sending the headers here as there are many responses that don't have status codes.
+            _pinnedHeaders = SerializeHeaders(pHttpResponse);
+        }
+
+        // From HttpSys, except does not write to response
+        private unsafe List<GCHandle> SerializeHeaders(HttpApiTypes.HTTP_RESPONSE_V2* pHttpResponse)
+        {
+            // TODO make the header collection readonly to stop modifications.
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/HttpProtocol.cs]: remove todo

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+
+            try
+            {
+                context = _application.CreateContext(this);
+
+                await _application.ProcessRequestAsync(context);
+                // TODO
+                //if (Volatile.Read(ref _requestAborted) == 0)
+                //{
+                //    VerifyResponseContentLength();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/IISHttpContextOfT.cs]: File a bug for this one.

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpRequestBody.cs [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+                        return (int)actual;
+                    }
+                    else if (result.IsCompleted)
+                    {
+                        return 0;
+                    }
+                }
+                finally
+                {
+                    _httpContext.Input.Reader.Advance(readableBuffer.End, readableBuffer.End);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/IISHttpRequestBody.cs]: What about 'actual'?

File: src/Microsoft.AspNetCore.Server.IIS/OutputProducer.cs [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+            ArraySegment<byte> buffer,
+            CancellationToken cancellationToken)
+        {
+            var writableBuffer = default(WritableBuffer);
+
+            lock (_contextLock)
+            {
+                if (_completed)
+                {
+                    return Task.CompletedTask;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/OutputProducer.cs]: ObjectDisposedException?

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpResponseBody.cs [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        public unsafe override void Write(byte[] buffer, int offset, int count)
+        {
+            WriteAsync(buffer, offset, count, CancellationToken.None).GetAwaiter().GetResult();
+        }
+
+        public override unsafe Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            return _httpContext.WriteAsync(new ArraySegment<byte>(buffer, offset, count), cancellationToken);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/IISHttpResponseBody.cs]: Why is the WriteAsync logic in HttpProtocol but the ReadAsync is in the read stream?

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpServer.cs [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+            // Start the server by registering the callback
+            // TODO the context may change here for shutdown.
+            NativeMethods.register_callbacks(_requestHandler, _shutdownHandler, (IntPtr)_httpServerHandle, (IntPtr)_httpServerHandle);
+
+            return Task.CompletedTask;
+        }
+
+        public Task StopAsync(CancellationToken cancellationToken)
+        {
+            // TODO: Drain pending requests
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/IISHttpServer.cs]: File issues for this

File: src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.Hosting.Abstractions" />
+    <PackageReference Include="Microsoft.AspNetCore.Http" />
+    <PackageReference Include="Microsoft.AspNetCore.Http.Extensions" />
+    <PackageReference Include="Microsoft.AspNetCore.HttpOverrides" />
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj]: You shouldn't need HttpOverrides or Kestrel.

File: test/IISIntegration.IISServerFunctionalTests/LargeResponseBodyTests.cs [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+{
+    public class LargeResponseBodyTests : LoggedTest
+    {
+
+        public LargeResponseBodyTests(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Theory]
+        [InlineData(10000, Skip = "See https://github.com/aspnet/IISIntegration/issues/424")]
```
Comment (Tratcher) [inline test/IISIntegration.IISServerFunctionalTests/LargeResponseBodyTests.cs]: You can put the Skip on the Theory, no?

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpRequestBody.cs (hunk 20 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/IISHttpRequestBody.cs]: Yeah I believe that should be advancing by 'actual' and not to the end of the buffer, right?

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpResponseBody.cs (hunk 22 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/IISHttpResponseBody.cs]: I'll file an issue for this.

Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/IISHttpResponseBody.cs]: https://github.com/aspnet/IISIntegration/issues/435

File: src/Microsoft.AspNetCore.Server.IIS/IISHttpServer.cs (hunk 23 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/IISHttpServer.cs]: https://github.com/aspnet/IISIntegration/issues/436

File: src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj (hunk 24 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj]: I think we need Kestrel for System.IO.Pipelines.

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj]: Those aren't available directly?

File: src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj [hunk 26]
```diff
@@ -0,0 +1,19 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+
+  <PropertyGroup>
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj]: You can condense this with the previous PropertyGroup block.

File: src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.Hosting.Abstractions" />
+    <PackageReference Include="Microsoft.AspNetCore.Http" />
+    <PackageReference Include="Microsoft.AspNetCore.Http.Extensions" />
+    <PackageReference Include="Microsoft.AspNetCore.Server.Kestrel" />
+    <PackageReference Include="Microsoft.AspNetCore.HttpSys.Sources" PrivateAssets="All" />
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj]: Nit: sort

File: test/IISIntegration.IISServerFunctionalTests/IISIntegration.IISServerFunctionalTests.csproj [hunk 28]
```diff
@@ -0,0 +1,25 @@
...
+  <ItemGroup>
+    <ProjectReference Include="..\..\src\Microsoft.AspNetCore.Server.IIS\Microsoft.AspNetCore.Server.IIS.csproj" />
+  </ItemGroup>
+  
+  <ItemGroup>
+    <Content Include="*.config" CopyToOutputDirectory="PreserveNewest" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" />
+    <PackageReference Include="xunit" />
```
Comment (natemcmaster) [inline test/IISIntegration.IISServerFunctionalTests/IISIntegration.IISServerFunctionalTests.csproj]: Nit: sort these lines.

File: src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj [hunk 29]
```diff
@@ -0,0 +1,19 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Server.IIS/Microsoft.AspNetCore.Server.IIS.csproj]: These are the other properties we define by default in shipping packages:

Description
PackageTags
GenerateDocumentationFile= true

Example:
https://github.com/aspnet/IISIntegration/blob/db84780e9369112d578c2733d2459a5972d15227/src/Microsoft.AspNetCore.Server.IISIntegration/Microsoft.AspNetCore.Server.IISIntegration.csproj#L4-L8

**Pull request merged**